draft-lenders-dns-over-coap.md

```diff
@@ -269,10 +269,18 @@ CoAP/CoRE Integration
 Proxies and caching
 -------------------
 
-TBD:
+After caching at one or more proxies, the Max-Age option in the CoAP header and the TTLs in the DNS
+message (set as specified in {{sec:resp-caching}}) may become out of sync. Both values are relative
+from the original send, but only the Max-Age option is updated at each intermediate cache.
+A DoC client MUST, as such, on reception of a DNS response check if the value of the Max-Age option
+in the CoAP header is less than the minimum TTL in the DNS message. In that case, it MUST subtract
+the difference of the minimum TTL and the value of the Max-Age option from all TTLs within the DNS
+response. All records for which resulting TTLs is equal to or less than 0, are to be considered to
+be out of date.
```

Comment:
> I find it weird that there are MUSTs here when the server only has a RECOMMENDED on setting Max-Age. If we allow the server to set any max-age, the calculations described here are not too meaningful.

Mh, my thinking for the MUST was here, that it needs to be done in order for the numbers to make sense. Maybe

```
If a Max-Age option is present in a response, a DoC client MUST, as such, on reception of a DNS response check if the value of the Max-Age option […]
```

> The underflow (less than 0) condition can not happen, because we do the subtraction $TTL_i - (min(TTL) - Max-Age)$, and with Max-Age being positive this never underflows.

Will fix.

> I'd prefer doing the subtraction on the server already. (That would allow using all records without any calculation as long as the resource is valid).

What is "the server" in that case? If it is the DoC server, the current spec already says that the DoC server should use the minimum TTL as value for the Max-Age option. The interesting thing is, what happens in-between at CoAP/REST proxies, that might be agnostic to DoC since they might be pure CoAP/REST and have nothing to do with name resolution. As such, we only have, _especially_ under the "REST, not CoAP" point of view, two points of interest where this can be tweaked: The DoC server or the DoC client. As the DoC server might not know if any proxies were used en-route, only the DoC client can make the difference here.

Reply:
> > I find it weird that there are MUSTs here when the server only has a RECOMMENDED on setting Max-Age. If we allow the server to set any max-age, the calculations described here are not too meaningful.
> 
> Mh, my thinking for the MUST was here, that it needs to be done in order for the numbers to make sense. Maybe
> 
> ```
> If a Max-Age option is present in a response, a DoC client MUST, as such, on reception of a DNS response check if the value of the Max-Age option […]
> ```
> 
> > The underflow (less than 0) condition can not happen, because we do the subtraction $TTL_i - (min(TTL) - Max-Age)$, and with Max-Age being positive this never underflows.
> 
> Will fix.

Done

Reply:
Replying from https://github.com/anr-bmbf-pivot/draft-dns-over-coap/pull/17#issuecomment-1059931619... GitHub's mail interface ripped this thread apart:

> On Sat, Mar 05, 2022 at 02:46:56AM -0800, Martine Lenders wrote:
> > ```
> > If a Max-Age option is present in a response, a DoC client MUST, as
> > such, on reception of a DNS response check if the value of the Max-Age
> > option […]
> > ```
> 
> The absence of Max-Age just means a Max-Age of 60 per
> <https://datatracker.ietf.org/doc/html/rfc7252#section-5.10.5>, so it's
> not really special.

Good hint! Will try to accommodate for that.

> > What is "the server" in that case? If it is the DoC server, the
> > current spec already says that the DoC server should use the minimum
> > TTL as value for the Max-Age option.
> 
> The entity that converts a DNS statement that is not encapsulated in
> something that has a lifetime of its own to a DNS statement that is
> encapsulated in something that may be used as-is for some time.
> 
> > The interesting thing is, what happens in-between at CoAP/REST
> > proxies, that might be agnostic to DoC since they might be pure
> > CoAP/REST and have nothing to do with name resolution. As such, we
> > only have, _especially_ under the "REST, not CoAP" point of view, two
> > points of interest where this can be tweaked: The DoC server or the
> > DoC client. As the DoC server might not know if any proxies were used
> > en-route, only the DoC client can make the difference here.
> 
> "might" ... proxies will in all likelihood be agnostic of DoC.
> 
> One of them should "know", but neither should need to. (The client knows
> of forward proxies, and the server or at least server operator knows of
> reverse proxies; none can exclude the presence of proxies on their own).
> 
> Proxies manage Max-Age, and while it might be justifiable for them to
> decrease it, they can't increase it. They both work no matter whether we
> precompute the difference or not, as long as all proxies are well
> behaved.
> 
> The advantage I see of precomputation is that the client that is
> interested in the TTLs (which would be the case if it maintains a
> DNS-style cache rather than a REST style cache) doesn't need to iterate
> over the result twice (once to find the minimum, another time to
> actually evaluate) but can go over it in a single pass. (It shifts the
> load to the DoC server that could otherwise iterate once and then
> memcpy, and now needs to iterate and subtract-while-copying). But it's
> only an implementation advantage; both styles have the good property
> that a client that needs to use a result "now" can just use any record
> in the response disregarding all timeouts.

Mh... after some offline discussion, I think I finally understand how you mean it. This is, however, a completely different direction to what I had in mind here. Will draft an alternative PR.

Reply:
> Replying from [#17 (comment)](https://github.com/anr-bmbf-pivot/draft-dns-over-coap/pull/17#issuecomment-1059931619)... GitHub's mail interface ripped this thread apart:
> 
> > On Sat, Mar 05, 2022 at 02:46:56AM -0800, Martine Lenders wrote:
> > > ```
> > > If a Max-Age option is present in a response, a DoC client MUST, as
> > > such, on reception of a DNS response check if the value of the Max-Age
> > > option […]
> > > ```
> > 
> > 
> > The absence of Max-Age just means a Max-Age of 60 per
> > https://datatracker.ietf.org/doc/html/rfc7252#section-5.10.5, so it's
> > not really special.
> 
> Good hint! Will try to accommodate for that.

Removed all mention of a Max-Age option being present or not and also tweaked the wording not to talk of the "value of the Max-Age option", but to say "the Max-Age [value] of the CoAP response".

Reply:
> Mh... after some offline discussion, I think I finally understand how you mean it. This is, however, a completely different direction to what I had in mind here. Will draft an alternative PR.

See #19